lib/aurora_bootstrapper/notifier.rb

```diff
@@ -0,0 +1,46 @@
+require 'aws-sdk-s3'
+
+module AuroraBootstrapper
+  class Notifier
+    def initialize( stub_responses: false )
+      region              = ENV.fetch( 'REGION', 'us-west-2')
+      @s3_client          = stub_responses ? Aws::S3::Client.new(stub_responses: true) : Aws::S3::Client.new(region: region)
+    end
+
+    def push_state?( export_date:, into_bucket: )
+
+      # remove the prefix s3://
+      into_bucket = into_bucket[5..-1]
+
+      # append export_date (if there is any) and empty state file DONE.txt
+      path = [into_bucket, export_date, 'DONE.txt' ].compact.join('/')
+      index = path.index('/')
+      bucket_name = path[0, index - 1]
+      object_key = path[index + 1..-1]
```

Comment:
am i reading this correctly that we take the bucket, append the date and done, split that back up by the first `/` into bucket name and object?

This is a little hard to track and test.  Can we split this into separate methods, like:
```ruby
class Notifier
  def initialize( s3_path: s3_path)
    @s3_path = s3_path
  end
  
  def notify
    client.put_object(
      bucket: bucket_name,
      key: object_key
    )
    AuroraBootstrapper.logger.info( message: "State file has been uploaded to S3 '#{bucket_name}/#{object_key}'." )
  rescue => e
    AuroraBootstrapper.logger.error( message: "State file failed to upload to S3 '#{bucket_name}/#{object_key}': #{e.messages}." )
  end

  protected

  def region
    @region ||= ENV.fetch( 'REGION', 'us-west-2' )
  end

  def client
    @client ||= Aws::S3::Client.new(region: region)
  end
  
  def bucket
    @bucket  ||=.unprefixed_path.split( '/' ).first
  end
  
  def object_key
    @object_key ||= [ bucket_path, export_date, filename ].join( '/' )
  end
  
  def export_date
    @export_date ||= ENV.fetch( 'EXPORT_DATE',
                                 DateTime.now.strftime("%Y-%m-%d") )
  end

  def filename
    'DONE.txt'
  end

  def unprefixed_path
    @unprefixed_path ||= @s3_path.gsub ("s3://", "" )
  end

  def bucket_path
    @bucket_path ||= ( unprefixed_path.split( '/' ) - [ bucket ] ).join( '/' )
  end
end
```

Reply:
export_date could return different values as we use now for each invocation. Still like my implementaion of initializaing export_date in the constructor of exporter

Reply:
That's fair. Should it be a property of the top level AuroraBootstrap module. Also: what happens when the export has no date?

Reply:
Move @export_date ||= ENV.fetch( 'EXPORT_DATE', DateTime.now.strftime("%Y-%m-%d") ) to initialize method; simply return @export_date in export_date method